### ISSUE TITLE
ffi.new() isn't JITted, when it should be

### ISSUE THREAD

**Issue by sergepetrenko** [tarantool/tarantool#4630]
We observed some strange decimal tuple field behaviour recently: accessing decimal tuple fields by index is slower than doing the same by name.
For example, run the following script: `time tarantool script.lua`

script.lua:
```lua
local decimal = require('decimal')
local num_iters = 1000000

box.cfg{}

if box.space.test ~= nil then box.space.test:drop() end 
s = box.schema.space.create('test', {format = { 
    { name = 'id', type = 'unsigned' },
    { name = 'dec', type = 'decimal' }
}})
s:create_index('pk')

tup = s:insert{0, decimal.new('1234.5678')}

for i = 1,num_iters do
    a = tup[2]
end

print(a)

os.exit()
```
The output will be:
```
1234.5678

real	0m8.155s
user	0m8.054s
sys	0m0.045s
```
However, if you change `a = tup[2]` to `a = tup.dec` in the loop, the output will be
```
1234.5678

real	0m0.697s
user	0m0.640s
sys	0m0.022s
```

Turns out the problem lies in luajit not JITting ffi.new("decimal_t") call.
jit.dump partial output is below
```
0000  . . . . FUNCF    7          ; msgpackffi.lua:497
0001  . . . . UGET     2   0      ; ffi
0002  . . . . TGETS    2   2   0  ; "new"
0003  . . . . KSTR     3   1      ; "decimal_t"
0004  . . . . CALL     2   2   2
0000  . . . . . FUNCC               ; ffi.new
```
...
```
0245 >  fun EQ     0244  ffi.new
0246 >  str EQ     "decimal_t"  "decimal_t"
0247 >  cdt CNEW   +106
0248    p64 ADD    0247  +8
0249    int XSTORE 0248  +0
0250    p64 ADD    0247  +12
0251    int XSTORE 0250  +0
0252    p64 ADD    0247  +16
0253    u8  XSTORE 0252  +0
---- TRACE 23 abort msgpackffi.lua:497 -- NYI: unsupported C type conversion
```
For the record, ffi's [NYI list](http://luajit.org/ext_ffi_semantics.html#status)  states, that the following operations are not JITted:
> Bitfield accesses and initializations.
> Vector operations.
> Table initializers.
> Initialization of nested struct/union types.
> Allocations of variable-length arrays or structs.
> Allocations of C types with a size > 128 bytes or an alignment > 8 bytes.
> Conversions from lightuserdata to void *.
> Pointer differences for element sizes that are not a power of two.
> Calls to C functions with aggregates passed or returned by value.
> Calls to ctype metamethods which are not plain functions.
> ctype __newindex tables and non-string lookups in ctype __index tables.
> tostring() for cdata types.
> Calls to ffi.cdef(), ffi.load() and ffi.metatype().

Here's our 'decimal_t' C declaration:
```c
typedef struct {
    int32_t digits;
    int32_t exponent;
    uint8_t bits;
    uint16_t lsu[13];
    } decimal_t;
```
Looks like it doesn't fall under any of the restrictions listed above, nevertheless FFI refuses to JIT compile
`ffi.new('decimal_t')`

**Comment by sergos** [comment]
Fix for #5885 fixes this one too

**Comment by igormunkin** [comment]
I re-run the benchmark[^1] before and after tarantool/luajit@ba8313c9. The results are below:
* Before (a85629a69):
```
$ ./src/tarantool -v
Tarantool 2.11.0-entrypoint-177-ga85629a69
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/bin/c++
C_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts  -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror
$ time ./src/tarantool ../4630-dec.lua 
1234.5678

real    0m0.476s
user    0m0.454s
sys     0m0.022s
$ time ./src/tarantool ../4630-2.lua 
1234.5678

real    0m6.751s
user    0m6.703s
sys     0m0.019s

```

* After (34330b159):
```
$ ./src/tarantool -v
Tarantool 2.11.0-entrypoint-178-g34330b159
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=ON
Compiler: /usr/bin/cc /usr/bin/c++
C_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -std=c11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -std=c++11 -Wall -Wextra -Wno-strict-aliasing -Wno-char-subscripts -Wno-format-truncation -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror
$ time ./src/tarantool ../4630-dec.lua 
1234.5678

real    0m0.508s
user    0m0.468s
sys     0m0.016s

$ time ./src/tarantool ../4630-2.lua 
1234.5678

real    0m0.402s
user    0m0.380s
sys     0m0.022s
```

So the issue is considered fixed since the submodule update in tarantool in 2.11.0-entrypoint-169-gd1d3d93ae (d1d3d93), 2.10.0-62-g54d798043 (54d7980), 1.10.13-27-geda8a1c36 (eda8a1c). Closing.

[^1]: I've created two files from the snippet above: `4630-dec.lua` with `tup.dec` lookup and `4630-2.lua` with `tup[2]` lookup.
